### ISSUE TITLE
Feedback and ideas for future

### ISSUE THREAD

**Issue by gre** [LedgerHQ/ledger-node-js-api#28]
assuming https://github.com/LedgerHQ/ledger-node-js-api/pull/26 gets in, here is next step feedback & ideas

## JS code

* The library have a lot of async. I think it would be interesting to migrate
  the promise codebase to **async/await** (babel compiles it down to Promise
  code) This is especially interesting for code that need to chain promises in
  sequence (`eachSeries`) (you can basically do a `for` loop with a `await` in
  it!).
* Apparently `new Buffer` is deprecated so need to port the remaining usage of
  that (and use `Buffer.from` / `Buffer.alloc`)
* codebase currently have some repetition. Need to think what we could factorize
  and if there is a higher level library that could help us on dealing with data
  format? (e.g; maybe we should directly manipulate the buffer, but define some
  "schema" and inject all data to "serialize" something) – just a thought.

## Documentation

* https://www.npmjs.com/package/documentation could be interesting to use. We
  could generate a API.md that is always in sync with the source code. What we
  need to do is to add some "Javadoc" in all public objects/classes/methods. To
  have used it before, I'm pretty sure this is compatible with Flow too. (need
  to check this works ok with async/await tho)
* README should start by telling more what the library is, and how to install
  and use it. The part about developing on the project is documented in
  CONTRIBUTING.md (and we can have a link to there).
* README is a good place to have a simple example so people instantly get the
  "gist" of the library. I'm not sure if the whole documentation should stick in
  there and maybe we would delegate to this API.md? (we have have links to that
  API doc)
* Contributing needs doc on how to deploy a new version on npm.

## API

These are breaking API changes suggestions (only to be in a major release)

* I think it would be better to directly export the classes (LedgerBtc /
  LedgerEth / ..) under these names.
* Maybe we should have LedgerComm exported as either `LedgerCommU2F` or
  `LedgerCommNode`. So we can just write the exact same code on Node or on
  Browser (and it also simplify documenting this).
* I don't think we should suffix any method with `_async`. It is common to have
  promise returned in JS libs. Just need users to know everything is async
  (RTFM!), but also if they use Flowtype, you won't have surprises ;)

More general question & thoughts

* This is unclear to me what is the lifecycle of a `comm` and with a
  LedgerBtc/LedgerEth instance. E.g. nor the API, nor any runtime check,
  prevents me from using comm on 2 instances of these (which I guess is an
  issue, especially that instanciating a LedgerBtc will set a state on comm
  (`setScrambleKey`)). Which led me to the question: can we afford creating one
  comm at a time & per LedgerBtc/Eth usage? If so, shouldn't this be hidden for
  the user? (aka you would do: `LedgerBtc.create().then(btc => ...)` without
  having to worry about comm ?). If we still want comm to have a bigger
  lifecycle (like you instanciate it only once for the all app lifecycle), I
  think we will need a "context" concept that we **bind** again that context
  when we enter a new call (e.g. like OpenGL programs) and maybe we even need
  some ways to throw if user is calling something and we're in the middle of
  something else (to avoid concurrent access).

## Tests

I suggest we could use Jest engine for tests.

* We could use test coverage to determine which things remain to be tested.
  (jest provides this)
* to test LedgerCommU2F, we could mock implementation of U2F and back it by
  node-hid. (jest have nice ways to .mock things). (we still would have browser
  example/test for a end-to-end check, but if the mock implementation is
  correct, we shouldn't worry much as it would just mean testing the actual
  browser U2F)
* tests can use async/await too (I think Jest works with async test too)
* This would still requires the user to plug a device and manually check&accept
  the things on the display. This is nice to have but I think we should also
  have a headless version (device simulator) that eventually runs on Travis. I
  imagine there will be some addition testing code for the simulator too, like
  you want to test both side (the client side and the "device" side) e.g; that
  it display something / a given string / ...

## CI

* as said above, if we can have a headless version (device simulator) we can
  eventually run tests on Travis!
* but also, we should already setup Travis check that will call the lint and
  flow commands (and obviously we connect it to incoming PRs).

**Comment by matthewdowney** [comment]
It would be great to be able to query an xpub/list of xpubs. Right now the solution I'm applying to get an xpub for some path (let's say  `p = m/49'/0'/0'/0/0`):

1. Get the public key & chain code at `p`.
2. Get the public key at the parent node of `p` (m/49'/0'/0'/0/).
3. Compress `p`'s public key, take `HASH160(parent_key)[:4]` as the fingerprint, & serialize the rest of the xpub from `p`'s information. 

I think it would also be intuitive to make a distinction between coin-specific functionality & general cryptographic functionality. For instance, the BIP32 public key derivation code is applicable for all secp256k1 coins. I understand that the ledger itself has separate apps for each coin type so this might not be possible to separate, but that would make more sense semantically if it's possible.

**Comment by gre** [comment]
most point addressed in https://github.com/LedgerHQ/ledger-node-js-api/pull/37 . internal refactoring with async/await will come in the future.

@matthewdowney please create a new issue if it is still a usecase dedicated to discussing on your feature request. BTW I am planning to refactor the  path parser using library https://github.com/axic/bip32-path

**Comment by matthewdowney** [comment]
Ok. I think it's a use case for anybody who'd like to use this library, but I've since moved on. If you'd like to implement it feel free—but it was just a suggestion so I'm not looking to have any ongoing role in its realization.
